cpp/server/handler.cc

```diff
@@ -237,38 +241,81 @@ int64_t GetIntParam(const multimap<string, string>& query,
 }  // namespace
 
 
-HttpHandler::HttpHandler(LogLookup<LoggedCertificate>* log_lookup,
-                         const ReadOnlyDatabase<LoggedCertificate>* db,
-                         const CertChecker* cert_checker, Frontend* frontend,
-                         ThreadPool* pool)
+HttpHandler::HttpHandler(
+    LogLookup<LoggedCertificate>* log_lookup,
+    const ReadOnlyDatabase<LoggedCertificate>* db,
+    const ClusterStateController<LoggedCertificate>* controller,
+    const CertChecker* cert_checker, Frontend* frontend, Proxy* proxy,
+    ThreadPool* pool)
     : log_lookup_(CHECK_NOTNULL(log_lookup)),
       db_(CHECK_NOTNULL(db)),
+      controller_(CHECK_NOTNULL(controller)),
       cert_checker_(CHECK_NOTNULL(cert_checker)),
       frontend_(frontend),
+      proxy_(CHECK_NOTNULL(proxy)),
       pool_(CHECK_NOTNULL(pool)) {
 }
 
 
+void HttpHandler::ProxyInterceptor(
+    const libevent::HttpServer::HandlerCallback& local_handler,
+    evhttp_request* request) {
+  // Need to punt this via the threadpool because the interaction with the
+  // controller can block pending other libevent updates, and since we're on
+  // the libevent thread here...
+  pool_->Add([this, request, local_handler]() {
+    VLOG(1) << "Running proxy interceptor...";
+    // TODO(alcutter): We can be a bit smarter about when to proxy off the
+    // request - being stale wrt to the current serving STH doesn't
+    // automatically mean we're unable to answer this request.
+    if (controller_->NodeIsStale()) {
+      const util::Status status(proxy_->ProxyRequest(request));
+      if (!status.ok()) {
+        LOG(INFO) << "Failed to proxy request: " << status;
+        SendError(request, HTTP_INTERNAL, "Failed to proxy request");
+      }
+      return;
+    } else {
+      local_handler(request);
```

Comment:
The inception of this PR predates that knowledge, and I forgot to update it after we learnt that last night :) (Note that the `Blocking...()` methods all originally executed on the pool too.)

Can't check `NodeIsStale()` before spinning off because that's on the libevent thread, and that method can block if the internal lock is held.  I had put a comment there about that, I've updated it to be even more explicit.

I've updated the Proxy to send the error message though, that's a good idea.